sim-administration/simmanager/src/main/kotlin/org/ostelco/simcards/admin/PollOutstandingProfilesTask.kt

```diff
@@ -0,0 +1,134 @@
+import org.ostelco.simcards.inventory.SmDpPlusState
+import org.ostelco.simcards.profilevendors.ProfileVendorAdapterFactory
+import java.io.PrintWriter
+
+/**
+ * A dropwizard "task" that is intended to be invoked as an administrative step
+ * by an external agent that is part of the serving system, not a customer of it.
+ *
+ * The task implements a task that for all of the sim profiles in storage
+ * will search for profiles that has been allocated to end user equipment
+ * but has not yet been downloaded or installed by it, as seen by ES2+
+ * callbacks.
+ *
+ * The functionality is duplicating some of the functions implemented by
+ * the ES2+ callback mechanism. The reason why we need this task in addition
+ * is that we have found the ES2+ callback to be unreliable.  This task
+ * is therefore intended to serve as a fallback for that when working correctly
+ * preferable mechanism.
+ */
+
+class PollOutstandingProfilesTask(
+        val simInventoryDAO: SimInventoryDAO,
+        val pvaf: ProfileVendorAdapterFactory) : Task("poll_outstanding_profiles") {
+
+    private val logger by getLogger()
+
+    @Throws(Exception::class)
+    override fun execute(parameters: ImmutableMultimap<String, String>, output: PrintWriter) {
+        // TODO: Rewrite to deliver a report of what happened, also if things went well.
+        pollAllocatedButNotDownloadedProfiles(output)
+                .mapLeft { simManagerError ->
+                    logger.error(simManagerError.description)
+                    output.println(asJson(simManagerError))
+                }
+    }
+
+    private fun pollAllocatedButNotDownloadedProfiles(output: PrintWriter): Either<SimManagerError, Unit> {
+        return simInventoryDAO.findAllocatedButNotDownloadedProfiles().map { profilesToPoll ->
+            pollForSmdpStatus(output, profilesToPoll)
+        }
+    }
+
+    private fun pollForSmdpStatus(output: PrintWriter, profilesToPoll: List<SimEntry>) {
+
+        val result = StringBuffer()
+
+        @Synchronized
+        fun reportln(s: String, e: Exception? = null) {
+            output.println(s)
+            if (e != null) {
+                output.println("   Caused exception: $e")
+            }
+        }
+
+        fun sendReport() {
+            output.print(result.toString())
+        }
+
+        fun updateProfileInDb(p: ProfileStatus) {
+            try {
+
+                // Get state and iccid non null strings
+                val stateString = p.state ?: ""
+                val iccid = p.iccid ?: ""
+
+                // Then discard empty states.
+                if (stateString == "") {
+                    reportln("Empty state detected for iccid ${p.iccid}")
+                    return
+                }
+
+                val state = SmDpPlusState.valueOf(stateString.toUpperCase())
+
+                if (state != SmDpPlusState.RELEASED) {
+                    val update = simInventoryDAO.setSmDpPlusStateUsingIccid(iccid, state)
```

Comment:
Rather than call the low-level DB functions (this calls `SimInventoryDB` directly) the API defined in `SimInventoryApi` should be called. If the required functionality is not there, it should be added.